src/test/java/org/datadog/jmxfetch/TestReconnectContainer.java

```diff
@@ -0,0 +1,200 @@
+import org.junit.Rule;
+import org.junit.Test;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.wait.strategy.Wait;
+import org.testcontainers.images.builder.ImageFromDockerfile;
+import org.testcontainers.utility.DockerImageName;
+
+import lombok.extern.slf4j.Slf4j;
+
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.file.Paths;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+
+import org.datadog.jmxfetch.reporter.ConsoleReporter;
+
+@Slf4j
+class JMXFailServerClient {
+    private final String host;
+    private final int port;
+
+    public JMXFailServerClient(String host, int port) {
+        this.host = host;
+        this.port = port;
+    }
+
+    public void startJMXServer() throws IOException {
+        sendPostRequest("/start");
+    }
+
+    public void stopJMXServer() throws IOException {
+        sendPostRequest("/stop");
+    }
+
+    public void jmxCutNetwork() throws IOException {
+        sendPostRequest("/jmx/cutNetwork");
+    }
+
+    public void jmxRestoreNetwork() throws IOException {
+        sendPostRequest("/jmx/restoreNetwork");
+    }
+
+    private void sendPostRequest(String endpoint) throws IOException {
+        URL url = new URL("http://" + host + ":" + port + endpoint);
+        log.info("Sending POST to {} ", url);
+        HttpURLConnection con = (HttpURLConnection) url.openConnection();
+        con.setRequestMethod("POST");
+
+        int responseCode = con.getResponseCode();
+
+        if (responseCode == HttpURLConnection.HTTP_OK) {
+            BufferedReader in = new BufferedReader(new InputStreamReader(con.getInputStream()));
+            String inputLine;
+            StringBuffer response = new StringBuffer();
+            while ((inputLine = in.readLine()) != null) {
+                response.append(inputLine);
+            }
+            in.close();
+            System.out.println(response.toString());
+        } else {
+            System.out.println("HTTP POST request failed with status code: " + responseCode);
+        }
+    }
+}
+
+
+public class TestReconnectContainer extends TestRemoteAppCommon {
+    private static final int rmiPort = 9090;
+    private static final int controlPort = 9091;
+
+    private static boolean isDomainPresent(String domain, MBeanServerConnection mbs) {
+        boolean found = false;
+        try {
+            String[] domains = mbs.getDomains();
+            for (int i = 0; i < domains.length; i++) {
+                if (domains[i].equals(domain)) {
+                    found = true;
+                }
+            }
+        } catch (IOException e) {
+            found = false;
+        }
+        return found;
+    }
+
+    private static ImageFromDockerfile img = new ImageFromDockerfile()
+        .withFileFromPath(".", Paths.get("./tools/misbehaving-jmx-server/"));
+
+    @Rule
+    public GenericContainer<?> cont = new GenericContainer<>(img)
```

Comment:
Add a "before"/"beforeEach" that configures the specific beans + domains that should be present for a given test.

Currently there is only 1 custom bean and it is hard-coded in `misbehavingjmxserver.App`